### PR TITLE
Update E2E tests, add new E2E tests and automatically run them on Jenkins

### LINKED ISSUE #66: Run integration tests for resdk on Travis

### PR DESCRIPTION
**Pull request by JenkoB**
Closes #66

### CODE REVIEW

**Comment by tjanez** [comment]
@dblenkus, I've finished my work on `Jenkinsfile`. Please, review it.
The only thing that is missing is publishing of the coverage report with the Cobertura Jenkins plugin. They've had a change of maintainership recently and the [new maintainer said he will attempt to merge support for Jenkins Pipeline in Cobertura this weekend](https://github.com/jenkinsci/cobertura-plugin/pull/55). But we can add that functionality later.

@dblenkus, @JenkoB, I've added CHANGELOG entries to your commits, please review them.

**Comment by tjanez** [comment]
The Travis CI fails due to the new version of Pylint. See #103.

File: resdk/tests/functional/base.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+"""Functiona tests for ReSDK."""
```
Comment (tjanez) [inline resdk/tests/functional/base.py]: Typo: "Functiona" -> "Functional"

File: resdk/tests/functional/analysis/e2e_alignment.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# pylint: disable=missing-docstring
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from resdk.tests.functional.base import BaseResdkFunctionalTest
+
+
+class TestAligners(BaseResdkFunctionalTest):
+
+    def test_bowtie2(self):
```
Comment (tjanez) [inline resdk/tests/functional/analysis/e2e_alignment.py]: Add note to the CHANGELOG about added E2E tests for `bowtie2` and `hisat2`.

File: resdk/tests/functional/analysis/e2e_chip_seq.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
+# pylint: disable=missing-docstring
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from resdk.tests.functional.base import BaseResdkFunctionalTest
+
+
+class TestChipSeq(BaseResdkFunctionalTest):
+
+    def test_macs(self):
```
Comment (tjanez) [inline resdk/tests/functional/analysis/e2e_chip_seq.py]: Same as above, add note to the CHANGELOG about added E2E tests for `macs` and `rose`.

File: resdk/analysis/chip_seq.py [hunk 4]
```diff
@@ -68,8 +68,10 @@ def macs(resource, use_background=True, p_value=None):
                 inputs['pvalue'] = p_value
 
             if use_background:
-                if is_background(sample):
-                    continue  # don't run process on the background sample
+                if is_background(sample) and not is_sample(single_resource):
```
Comment (tjanez) [inline resdk/analysis/chip_seq.py]: Would this change/fix be worth mentioning in the commit message?

File: resdk/shortcuts/collection.py [hunk 5]
```diff
@@ -65,7 +65,7 @@ def _create_relation(self, relation_type, samples, positions=[], label=None):
         if label:
             relation_data['label'] = label
 
-        self.resolwe.relation.create(**relation_data)
+        return self.resolwe.relation.create(**relation_data)
```
Comment (tjanez) [inline resdk/shortcuts/collection.py]: Similar as above, would this change/fix be worth mentioning in the commit message?

File: tox.ini [hunk 6]
```diff
@@ -33,6 +33,8 @@ commands=
     pycodestyle resdk
 # check PEP 257
     pydocstyle resdk
+# check order of imports
```
Comment (tjanez) [inline tox.ini]: Nice 🙂 !

File: resdk/tests/functional/base.py [hunk 7]
```diff
@@ -118,3 +134,30 @@ def get_macs(self, count=1, collection=None):
...
+        for _ in range(count):
+            cuffquant = self.res.run(
+                'upload-cxb',
+                input={'src': cuffquant_path, 'source': 'NCBI'},
+                **collections
+            )
+
+            cuffquants.append(cuffquant)
+
+            # TODO: Remove this
```
Comment (tjanez) [inline resdk/tests/functional/base.py]: Hmm... does this mean the code below should be removed?

**Comment by dblenkus** [comment]
@tjanez I've addressed all your comments. But I think that changelog should contain only changes that are important from user's point of view. So I would remove all entries connected to adding/updating/running tests.

**Comment by tjanez** [comment]
> @tjanez I've addressed all your comments. 

Great, thanks!

> But I think that changelog should contain only changes that are important from user's point of view. So I would remove all entries connected to adding/updating/running tests.

Well, the CHANGELOG is also useful for developers of the project. I've picked some high-profile projects and looked at how they treat tests in their CHANGELOG/Release notes:
- [pip's Release notes](https://pip.pypa.io/en/stable/news/) mentions things like:
  - Add Appveyor CI
  - Moved tests to be run on Travis CI: http://travis-ci.org/pypa/pip
  - Significant test framework changes, from Hugo Lopes Tavares
- [Ansible's CHANGELOG](https://github.com/ansible/ansible/blob/devel/CHANGELOG.md) has nice separation between **major changes** and **minor changes**. For example, under minor changes, they mention:
  - Many more tests. The new API makes things more testable and we took advantage of it.
  
  And they mention these under major changes:
  - Added travis integration to github for basic tests, this should speed up ticket triage and merging.
  - Start of new integration test infrastructure
- [Django's release notes](https://docs.djangoproject.com/en/1.11/releases/1.11/) is a well thought-out document on its own. They list the most important changes first and then group all minor features under **Minor features**, but still separate them by the area they touch. I haven't found anything mentioning changes to their test infrastructure and tests (besides the test facitilities Django itself provides to its applications).

My opinion is that some changes (especially major ones) regarding how the tests are run and when the test framework is substantially updated, should be part of the CHANGELOG. I also see that it would be useful to separate user-visible and developer-visible changes to separate sections in the future, so that users don't need to be bothered by developer-visible changes.

@dblenkus, what do you think?

**Comment by dblenkus** [comment]
@tjanez As you said, we can include major changes (i.e. update E2E tests and run them on Jenkins), but I would definitely avoid listing all added tests. But I don't think that this project is big enough (or is developing rapidly enough) to make sense to have separate changelog for developers.

**Comment by tjanez** [comment]
> @tjanez As you said, we can include major changes (i.e. update E2E tests and run them on Jenkins), but I would definitely avoid listing all added tests.

Agreed.

> But I don't think that this project is big enough (or is developing rapidly enough) to make sense to have separate changelog for developers.

Yes, agreed.

@dblenkus, just fix commit 1f8834919df5d09c3d86ac9dea92761bb240c713 and:

  - remove _E2E tests for ``cuffquant`` and ``cuffnorm``_ CHANGELOG entry,
  - move the _``isort`` linter_ CHANGELOG entry to the _Add isort linter_ commit.

Then this is ready to be merged.

**Pull request merged**